File: mobile/modules/spotify-app-remote/android/src/main/java/expo/modules/spotifyappremote/SpotifyAppRemoteModule.kt
package expo.modules.spotifyappremote

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.Promise
import com.spotify.android.appremote.api.SpotifyAppRemote
import com.spotify.android.appremote.api.ConnectionParams
import com.spotify.android.appremote.api.Connector
import com.spotify.protocol.types.PlayerState

const val CLIENT_ID = "18f1b52ab93b4c6480b1599b64d9be5b"
const val REDIRECT_URI = "partytime://callback"

class SpotifyAppRemoteModule : Module() {
  private var spotifyAppRemote: SpotifyAppRemote? = null
  private var isSubscribed = false

  override fun definition() = ModuleDefinition {
    Name("ExpoSpotifyAppRemote")

    Events("onPlayerStateChanged", "onConnectionChanged")

    AsyncFunction("connect") { accessToken: String, promise: Promise ->
      val activity = appContext.currentActivity
        ?: return@AsyncFunction promise.reject("NO_ACTIVITY", "No activity available", null)

      val params = ConnectionParams.Builder(CLIENT_ID)
        .setRedirectUri(REDIRECT_URI)
        .showAuthView(false)
        .build()

      SpotifyAppRemote.connect(activity, params, object : Connector.ConnectionListener {
        override fun onConnected(remote: SpotifyAppRemote) {
          spotifyAppRemote = remote
          sendEvent("onConnectionChanged", mapOf("connected" to true))
          promise.resolve(null)
        }

        override fun onFailure(error: Throwable) {
          sendEvent("onConnectionChanged", mapOf("connected" to false, "error" to (error.message ?: "Unknown error")))
          promise.reject("CONNECTION_FAILED", error.message ?: "Connection failed", error)
        }
      })
    }

    AsyncFunction("disconnect") { promise: Promise ->
      spotifyAppRemote?.let {
        SpotifyAppRemote.disconnect(it)
      }
      spotifyAppRemote = null
      isSubscribed = false
      promise.resolve(null)
    }

    AsyncFunction("play") { uri: String, promise: Promise ->
      val remote = spotifyAppRemote
        ?: return@AsyncFunction promise.reject("NOT_CONNECTED", "Not connected to Spotify", null)

      remote.playerApi.play(uri).setResultCallback {
        promise.resolve(null)
      }.setErrorCallback { error ->
        promise.reject("PLAY_ERROR", error.message ?: "Play failed", null)
      }
    }

    AsyncFunction("pause") { promise: Promise ->
      val remote = spotifyAppRemote
        ?: return@AsyncFunction promise.reject("NOT_CONNECTED", "Not connected to Spotify", null)

      remote.playerApi.pause().setResultCallback {
        promise.resolve(null)
      }.setErrorCallback { error ->
        promise.reject("PAUSE_ERROR", error.message ?: "Pause failed", null)
      }
    }

    AsyncFunction("resume") { promise: Promise ->
      val remote = spotifyAppRemote
        ?: return@AsyncFunction promise.reject("NOT_CONNECTED", "Not connected to Spotify", null)

      remote.playerApi.resume().setResultCallback {
        promise.resolve(null)
      }.setErrorCallback { error ->
        promise.reject("RESUME_ERROR", error.message ?: "Resume failed", null)
      }
    }

    AsyncFunction("seekTo") { positionMs: Int, promise: Promise ->
      val remote = spotifyAppRemote
        ?: return@AsyncFunction promise.reject("NOT_CONNECTED", "Not connected to Spotify", null)

      remote.playerApi.seekTo(positionMs.toLong()).setResultCallback {
        promise.resolve(null)
      }.setErrorCallback { error ->
        promise.reject("SEEK_ERROR", error.message ?: "Seek failed", null)
      }
    }

    AsyncFunction("skipToNext") { promise: Promise ->
      val remote = spotifyAppRemote
        ?: return@AsyncFunction promise.reject("NOT_CONNECTED", "Not connected to Spotify", null)

      remote.playerApi.skipNext().setResultCallback {
        promise.resolve(null)
      }.setErrorCallback { error ->
        promise.reject("SKIP_ERROR", error.message ?: "Skip failed", null)
      }
    }

    AsyncFunction("getPlayerState") { promise: Promise ->
      val remote = spotifyAppRemote
        ?: return@AsyncFunction promise.reject("NOT_CONNECTED", "Not connected to Spotify", null)

      remote.playerApi.playerState.setResultCallback { state ->
        promise.resolve(stateToMap(state))
      }.setErrorCallback { error ->
        promise.reject("STATE_ERROR", error.message ?: "Failed to get state", null)
      }
    }

    AsyncFunction("subscribeToPlayerState") { promise: Promise ->
      val remote = spotifyAppRemote
        ?: return@AsyncFunction promise.reject("NOT_CONNECTED", "Not connected to Spotify", null)

      remote.playerApi.subscribeToPlayerState().setEventCallback { state ->
        sendEvent("onPlayerStateChanged", stateToMap(state))
      }.setErrorCallback { error ->
        promise.reject("SUBSCRIBE_ERROR", error.message ?: "Subscribe failed", null)
      }
      isSubscribed = true
      promise.resolve(null)
    }

    AsyncFunction("unsubscribeFromPlayerState") { promise: Promise ->
      isSubscribed = false
      promise.resolve(null)
    }
  }

  private fun stateToMap(state: PlayerState): Map<String, Any?> {
    return mapOf(
      "uri" to (state.track?.uri ?: ""),
      "trackName" to (state.track?.name ?: ""),
      "artistName" to (state.track?.artist?.name ?: ""),
      "albumName" to (state.track?.album?.name ?: ""),
      "durationMs" to (state.track?.duration ?: 0L),
      "positionMs" to state.playbackPosition,
      "isPaused" to state.isPaused,
    )
  }
}

File: mobile/modules/spotify-app-remote/ios/SpotifyAppRemoteModule.swift
import ExpoModulesCore
import SpotifyiOS

fileprivate let CLIENT_ID = "18f1b52ab93b4c6480b1599b64d9be5b"
fileprivate let REDIRECT_URI = "partytime://callback"

public class SpotifyAppRemoteModule: Module {
  fileprivate var config: SPTConfiguration?
  fileprivate var appRemote: SPTAppRemote?
  fileprivate var delegateHandler: SpotifyDelegateHandler?
  fileprivate var connectPromise: Promise?
  fileprivate var isSubscribed = false

  fileprivate func createAppRemote(accessToken: String) {
    // Always destroy previous instance to avoid stale IPC state
    if let old = appRemote { old.disconnect() }
    config = SPTConfiguration(clientID: CLIENT_ID, redirectURL: URL(string: REDIRECT_URI)!)
    config!.playURI = ""
    appRemote = SPTAppRemote(configuration: config!, logLevel: .debug)
    delegateHandler = SpotifyDelegateHandler(module: self)
    appRemote!.delegate = delegateHandler
    appRemote!.connectionParameters.accessToken = accessToken
  }

  public func definition() -> ModuleDefinition {
    Name("ExpoSpotifyAppRemote")

    Events("onPlayerStateChanged", "onConnectionChanged")

    AsyncFunction("connect") { (accessToken: String, promise: Promise) in
      // Check Spotify is installed
      if let url = URL(string: "spotify:"), !UIApplication.shared.canOpenURL(url) {
        promise.reject("SPOTIFY_NOT_INSTALLED", "Spotify app is not installed")
        return
      }

      self.createAppRemote(accessToken: accessToken)
      self.connectPromise = promise

      self.appRemote!.connect()

      // 6-second timeout — SPTAppRemote can hang silently
      DispatchQueue.main.asyncAfter(deadline: .now() + 6.0) {
        if let p = self.connectPromise {
          self.connectPromise = nil
          p.reject("CONNECT_TIMEOUT", "Spotify connect timed out — is Spotify open?")
        }
      }
    }

    AsyncFunction("authorize") { (uri: String, promise: Promise) in
      guard let appRemote = self.appRemote else {
        promise.reject("NOT_INITIALIZED", "Call connect first to initialize App Remote")
        return
      }
      self.connectPromise = promise
      appRemote.authorizeAndPlayURI(uri.isEmpty ? "" : uri)
    }

    AsyncFunction("handleAuthURL") { (urlString: String, promise: Promise) in
      guard let url = URL(string: urlString),
            let appRemote = self.appRemote else {
        promise.resolve(false)
        return
      }
      let params = appRemote.authorizationParameters(from: url)
      if let token = params?[SPTAppRemoteAccessTokenKey] as? String {
        appRemote.connectionParameters.accessToken = token
        self.connectPromise = promise
        appRemote.connect()
      } else if let errorDesc = params?[SPTAppRemoteErrorDescriptionKey] as? String {
        promise.reject("AUTH_ERROR", errorDesc)
      } else {
        promise.resolve(false)
      }
    }

    AsyncFunction("disconnect") { (promise: Promise) in
      self.appRemote?.disconnect()
      self.appRemote = nil
      self.config = nil
      self.delegateHandler = nil
      self.isSubscribed = false
      promise.resolve(nil)
    }

    AsyncFunction("play") { (uri: String, promise: Promise) in
      guard let playerAPI = self.appRemote?.playerAPI else {
        promise.reject("NOT_CONNECTED", "Not connected to Spotify")
        return
      }
      playerAPI.play(uri) { _, error in
        if let error = error {
          promise.reject("PLAY_ERROR", error.localizedDescription)
        } else {
          promise.resolve(nil)
        }
      }
    }

    AsyncFunction("pause") { (promise: Promise) in
      guard let playerAPI = self.appRemote?.playerAPI else {
        promise.reject("NOT_CONNECTED", "Not connected to Spotify")
        return
      }
      playerAPI.pause { _, error in
        if let error = error {
          promise.reject("PAUSE_ERROR", error.localizedDescription)
        } else {
          promise.resolve(nil)
        }
      }
    }

    AsyncFunction("resume") { (promise: Promise) in
      guard let playerAPI = self.appRemote?.playerAPI else {
        promise.reject("NOT_CONNECTED", "Not connected to Spotify")
        return
      }
      playerAPI.resume { _, error in
        if let error = error {
          promise.reject("RESUME_ERROR", error.localizedDescription)
        } else {
          promise.resolve(nil)
        }
      }
    }

    AsyncFunction("seekTo") { (positionMs: Int, promise: Promise) in
      guard let playerAPI = self.appRemote?.playerAPI else {
        promise.reject("NOT_CONNECTED", "Not connected to Spotify")
        return
      }
      playerAPI.seek(toPosition: positionMs) { _, error in
        if let error = error {
          promise.reject("SEEK_ERROR", error.localizedDescription)
        } else {
          promise.resolve(nil)
        }
      }
    }

    AsyncFunction("skipToNext") { (promise: Promise) in
      guard let playerAPI = self.appRemote?.playerAPI else {
        promise.reject("NOT_CONNECTED", "Not connected to Spotify")
        return
      }
      playerAPI.skip(toNext: { _, error in
        if let error = error {
          promise.reject("SKIP_ERROR", error.localizedDescription)
        } else {
          promise.resolve(nil)
        }
      })
    }

    AsyncFunction("getPlayerState") { (promise: Promise) in
      guard let playerAPI = self.appRemote?.playerAPI else {
        promise.reject("NOT_CONNECTED", "Not connected to Spotify")
        return
      }
      playerAPI.getPlayerState { result, error in
        if let error = error {
          promise.reject("STATE_ERROR", error.localizedDescription)
          return
        }
        guard let state = result as? SPTAppRemotePlayerState else {
          promise.reject("STATE_ERROR", "Invalid player state")
          return
        }
        promise.resolve(self.stateToDict(state))
      }
    }

    AsyncFunction("subscribeToPlayerState") { (promise: Promise) in
      guard let playerAPI = self.appRemote?.playerAPI else {
        promise.reject("NOT_CONNECTED", "Not connected to Spotify")
        return
      }
      playerAPI.delegate = self.delegateHandler
      playerAPI.subscribe(toPlayerState: { _, error in
        if let error = error {
          promise.reject("SUBSCRIBE_ERROR", error.localizedDescription)
        } else {
          self.isSubscribed = true
          promise.resolve(nil)
        }
      })
    }

    AsyncFunction("unsubscribeFromPlayerState") { (promise: Promise) in
      guard let playerAPI = self.appRemote?.playerAPI else {
        promise.resolve(nil)
        return
      }
      playerAPI.unsubscribe(toPlayerState: { _, error in
        self.isSubscribed = false
        promise.resolve(nil)
      })
    }
  }

  fileprivate func stateToDict(_ state: SPTAppRemotePlayerState) -> [String: Any] {
    return [
      "uri": state.track.uri,
      "trackName": state.track.name,
      "artistName": state.track.artist.name,
      "albumName": state.track.album.name,
      "durationMs": state.track.duration,
      "positionMs": state.playbackPosition,
      "isPaused": state.isPaused,
    ]
  }

  fileprivate func handleConnect() {
    connectPromise?.resolve(nil)
    connectPromise = nil
    sendEvent("onConnectionChanged", ["connected": true])
  }

  fileprivate func handleConnectionFailed(_ error: Error) {
    connectPromise?.reject("CONNECTION_FAILED", error.localizedDescription)
    connectPromise = nil
    sendEvent("onConnectionChanged", ["connected": false, "error": error.localizedDescription])
  }

  fileprivate func handleDisconnect(_ error: Error?) {
    sendEvent("onConnectionChanged", ["connected": false, "error": error?.localizedDescription ?? ""])
  }

  fileprivate func handlePlayerStateChange(_ state: SPTAppRemotePlayerState) {
    sendEvent("onPlayerStateChanged", stateToDict(state))
  }
}

// NSObject proxy — SPTAppRemoteDelegate requires NSObjectProtocol conformance,
// but Expo's Module base class is not an NSObject subclass.
fileprivate class SpotifyDelegateHandler: NSObject, SPTAppRemoteDelegate, SPTAppRemotePlayerStateDelegate {
  weak var module: SpotifyAppRemoteModule?

  init(module: SpotifyAppRemoteModule) {
    self.module = module
    super.init()
  }

  func appRemoteDidEstablishConnection(_ appRemote: SPTAppRemote) {
    module?.handleConnect()
  }

  func appRemote(_ appRemote: SPTAppRemote, didFailConnectionAttemptWithError error: Error?) {
    module?.handleConnectionFailed(error ?? NSError(domain: "SpotifyAppRemote", code: -1, userInfo: [NSLocalizedDescriptionKey: "Unknown connection error"]))
  }

  func appRemote(_ appRemote: SPTAppRemote, didDisconnectWithError error: Error?) {
    module?.handleDisconnect(error)
  }

  func playerStateDidChange(_ playerState: SPTAppRemotePlayerState) {
    module?.handlePlayerStateChange(playerState)
  }
}
